From 55a699497aa983856e76f2e56ca4103a024bb7d6 Mon Sep 17 00:00:00 2001
From: Marco Kawajiri <kawajiri@meta.com>
Date: Wed, 27 Mar 2024 09:54:11 -0700
Subject: [PATCH] user-mgr: Add max_system_users as a build-time config

Added build-time option to configure max number of system users.
Defaults to 30 (same as previous hard-coded limit)

Tested:
State of system already exceeding max number of system users,
before:
```
$ journalctl -f
... phosphor-user-manager[10369]: Non-ipmi User limit reached
... phosphor-user-manager[10369]: No resource available.

$ curl $bmc/redfish/v1/AccountService/Accounts | jq -Sc .Members[] \
| wc -l
16
```

After:
```
$ curl $bmc/redfish/v1/AccountService/Accounts | jq -Sc .Members[] \
| wc -l
19
```

Change-Id: I8201d58c4f5051a35f43776e25fa90d7eef1856b
Signed-off-by: Marco Kawajiri <kawajiri@meta.com>
---
 meson.build   | 6 ++++++
 meson.options | 7 +++++++
 user_mgr.hpp  | 3 ++-
 3 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 1efe9d2..54370fd 100644
--- a/meson.build
+++ b/meson.build
@@ -23,6 +23,12 @@ conf_data.set_quoted(
     description: 'The DBus busname to own.',
 )
 
+conf_data.set(
+    'MAX_SYSTEM_USERS',
+    get_option('max_system_users'),
+    description: 'Maximum allowed number of system users',
+)
+
 conf_data.set(
     'CLASS_VERSION',
     1,
diff --git a/meson.options b/meson.options
index 95df809..038a42b 100644
--- a/meson.options
+++ b/meson.options
@@ -15,3 +15,10 @@ option(
     value: 64,
     description: 'Max password Length default value as per NIST guideline recommendation',
 )
+
+option(
+    'max_system_users',
+    type: 'integer',
+    value: 30,
+    description: 'Maximum allowed number of system users',
+)
diff --git a/user_mgr.hpp b/user_mgr.hpp
index 3aff46e..07f8811 100644
--- a/user_mgr.hpp
+++ b/user_mgr.hpp
@@ -14,6 +14,7 @@
 // limitations under the License.
 */
 #pragma once
+#include "config.h"
 #include "json_serializer.hpp"
 #include "users.hpp"
 
@@ -58,7 +59,7 @@ struct SystemUserInfo
 };
 
 inline constexpr size_t ipmiMaxUsers = 15;
-inline constexpr size_t maxSystemUsers = 30;
+inline constexpr size_t maxSystemUsers = MAX_SYSTEM_USERS;
 inline constexpr uint8_t minPasswdLength = 8;
 extern uint8_t maxPasswdLength; // MAX_PASSWORD_LENGTH;
 inline constexpr size_t maxSystemGroupNameLength = 32;
-- 
2.52.0

